src/runtime/runtime_unix.go

```diff
@@ -307,3 +326,164 @@ func growHeap() bool {
+	if s >= 32 {
+		// TODO: to support higher signal numbers, we need to turn
+		// receivedSignals into a uint32 array.
+		runtimePanicAt(returnAddress(0), "unsupported signal number")
+	}
+	tinygo_signal_ignore(s)
+}
+
+//go:linkname signal_disable os/signal.signal_disable
+func signal_disable(s uint32) {
+	if s >= 32 {
+		// TODO: to support higher signal numbers, we need to turn
+		// receivedSignals into a uint32 array.
+		runtimePanicAt(returnAddress(0), "unsupported signal number")
+	}
+	tinygo_signal_disable(s)
+}
+
+//go:linkname signal_waitUntilIdle os/signal.signalWaitUntilIdle
+func signal_waitUntilIdle() {
+	// Make sure all signals are sent on the channel.
+	for atomic.LoadUint32(&receivedSignals) != 0 {
+		checkSignals()
+		Gosched()
+	}
+
+	// Make sure all signals are processed.
+	for len(signalChan) != 0 {
+		Gosched()
+	}
+}
+
+//export tinygo_signal_enable
+func tinygo_signal_enable(s uint32)
+
+//export tinygo_signal_ignore
+func tinygo_signal_ignore(s uint32)
+
+//export tinygo_signal_disable
+func tinygo_signal_disable(s uint32)
+
+// void tinygo_signal_handler(int sig);
+//
+//export tinygo_signal_handler
+func tinygo_signal_handler(s int32) {
+	// This loop is essentially the atomic equivalent of the following:
+	//
+	//   receivedSignals |= 1 << s
+	//
+	// TODO: use atomic.Uint32.And once we drop support for Go 1.22 instead of
+	// this loop.
+	for {
+		mask := uint32(1) << uint32(s)
+		val := atomic.LoadUint32(&receivedSignals)
+		swapped := atomic.CompareAndSwapUint32(&receivedSignals, val, val|mask)
+		if swapped {
+			break
+		}
+	}
+}
+
+//go:linkname signal_recv os/signal.signal_recv
+func signal_recv() uint32 {
+	// Function called from os/signal to get the next received signal.
+	val := <-signalChan
+	checkSignals()
+	return val
+}
+
+// Atomically find a signal that previously occured and send it into the
+// signalChan channel. Return true if at least one signal was delivered this
+// way, false otherwise.
+func checkSignals() bool {
+	gotSignals := false
+	for {
+		// Extract the lowest numbered signal number from receivedSignals.
+		val := atomic.LoadUint32(&receivedSignals)
+		if val == 0 {
+			// There is no signal ready to be received by the program (common
+			// case).
+			return gotSignals
+		}
+		num := uint32(bits.TrailingZeros32(val))
+
+		// Do a non-blocking send on signalChan.
+		select {
+		case signalChan <- num:
+			// There was room free in the channel, so remove the signal number
+			// from the receivedSignals mask.
+			gotSignals = true
+		default:
+			// Could not send the signal number on the channel. This means
+			// there's still a signal pending. In that case, let it be received
+			// at which point checkSignals is called again to put the next one
+			// in the channel buffer.
+			return gotSignals
+		}
+
+		// Atomically clear the signal number from receivedSignals.
+		// TODO: use atomic.Uint32.Or once we drop support for Go 1.22 instead
+		// of this loop.
```

Comment:
Since Go 1.22 support will not be dropped in the next view releases, we should also not consider this a blocking factor here.

Reply:
Correct, all it will do is make the code look better (and _maybe_ a bit faster though that's probably not noticeable). I intentionally put in "Go 1.22" because when dropping support for a Go version I typically will grep for such strings to fix cases like these.